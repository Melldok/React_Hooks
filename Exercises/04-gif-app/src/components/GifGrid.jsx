
import { useFetchGifs } from "../hooks/useFetchGifs"
import { GifCard } from "./GifCard"
import { PropTypes } from "prop-types";




export const GifGrid = ({ category }) => {  

    const {images, isLoading} = useFetchGifs( category );
  
    return (
        <>
            <h3>{ category }</h3>
            {isLoading && <h2>Loading...</h2>}

            <div className="card-grid">
                {
                    images.map((image) => (
                        <GifCard 
                            key={image.id}
                            {...image}
                            // Spread operator can be used to spread all the props of the component
                            />
                    ))
                }
            </div>
            
        </>
  )
}


GifGrid.propTypes = {
    category : PropTypes.string.isRequired,
}