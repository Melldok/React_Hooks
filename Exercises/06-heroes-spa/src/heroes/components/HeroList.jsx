import { getHeroesByPublisher } from "../helpers"
import { HeroCard } from "./HeroCard";

export const HeroList = ({ publisher }) => {

    // helper function that filters the data array by publisher (Sent by props)
    const heroes = getHeroesByPublisher(publisher);

  return (
    <div className="row rows-cols-1 row-cols-md-3 g-3">
        {
            heroes.map(hero => (
               <HeroCard key={ hero.id } 
               // spread operator used to add all the props from the item, instead of writing them one by one
               {...hero}
               /> 
            ))
        }
    </div>
  )
}
