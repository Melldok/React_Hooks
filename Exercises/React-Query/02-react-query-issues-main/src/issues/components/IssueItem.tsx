import { FiInfo, FiMessageSquare, FiCheckCircle } from 'react-icons/fi';
import { Issue } from '../interfaces';
import { FC } from 'react';
import { useNavigate } from 'react-router-dom';
import { useQueryClient } from '@tanstack/react-query';
import { getIssueComments, getIssueInfo } from '../hooks/useIssue';
import { timeSince } from '../../helpers/timeSince';

interface Props {
    issue: Issue
}


export const IssueItem: FC<Props> = ({issue}) => {



    const navigate = useNavigate();
    const queryClient = useQueryClient();

    const prefetchData = () => {
        
        // Prefetch nos sirve para precargar información de una query antes de que el usuario la necesite. Esto es útil si por ejemplo sabemos que el usuario va a necesitar la información de un item en concreto, y queremos que cuando el usuario haga click en el item, la información ya esté cargada, esto se puede lanzar con diferentes eventos, como por ejemplo onMouseEnter.\

        queryClient.prefetchQuery(
            ['issue', issue.number],
            () => getIssueInfo(issue.number),
            
        
        ) 
        queryClient.prefetchQuery(
            ['issue', issue.number, 'comments'],
            () => getIssueComments(issue.number),
            
        
        ) 
    }

    const presetData = () => {
        
        // Preset data actúa de la misma manera que prefetch data, pero solo lo carga una vez
        queryClient.setQueryData(
            ['issue', issue.number],
            issue,
            {
                // Aquí podemos especificar el tiempo que queremos que se mantenga la información en cache
                updatedAt: new Date().getTime() + 10000
            }
            
        
        ) 
    }



    return (
        <div className="card mb-2 issue"
            onClick={() => navigate(`/issues/issue/${issue.number}`)}
            onMouseEnter={presetData}
        >
            <div className="card-body d-flex align-items-center">

                {
                    issue.state === 'open' 
                    ? <FiInfo size={30} color='red' />
                    : <FiCheckCircle size={30} color='green' />
                }
                
                

                <div className="d-flex flex-column flex-fill px-2">
                    <span>{issue.title}</span>
                    <span className="issue-subinfo">#{issue.number} opened {timeSince(issue.created_at)} <span className='fw-bold'>{issue.user.login}</span></span>
                    <div>
                        {
                            issue.labels.map(label => (
                                <span
                                    key={label.id}
                                    className="badge rounded-pill m-1"
                                    style={{ backgroundColor: `#${label.color}`, color : 'black' }}

                                >
                                    {label.name}

                                </span>
                            ))
                        }
                    </div>
                </div>

                <div className='d-flex align-items-center'>
                    <img src={issue.user.avatar_url} className="avatar" />
                    <span className='px-2'>{issue.comments}</span>
                    <FiMessageSquare />
                </div>

            </div>
        </div>
    )
}
