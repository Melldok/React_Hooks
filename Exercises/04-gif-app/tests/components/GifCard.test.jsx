import { render, screen } from "@testing-library/react"
import { GifCard } from "../../src/components/GifCard"




describe('Testing on GifCard', () => { 

    const title = 'Saitama';
    const url = 'https://one-punch.com/saitama.jpg'
    
    test('Should match snapshot', () => { 
        
        const {container} = render(<GifCard title={title} url={url} id="2"/>);
        expect(container).toMatchSnapshot();

     })

    test('Should show image with indicated URL and ALT', () => { 
        render(<GifCard title={title} url={url}/>)
        //screen.debug()
        // expect(screen.getByRole('img').src).toBe(url)
        // expect(screen.getByRole('img').alt).toBe(title)

        const {src, alt} = screen.getByRole('img')
        expect(src).toBe(url);
        expect(alt).toBe(title)

      })

      test('should show title on component', () => { 
        render(<GifCard title={title} url={url}/>)
        expect(screen.getByText(title)).toBeTruthy()

        
       })
 })