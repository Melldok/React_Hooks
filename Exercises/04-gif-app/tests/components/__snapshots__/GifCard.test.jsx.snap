// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing on GifCard Should match snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="Saitama"
      src="https://one-punch.com/saitama.jpg"
    />
    <p>
      Saitama
    </p>
  </div>
</div>
`;
